src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_fallback_formula_metric_fetcher.py

```diff
@@ -0,0 +1,54 @@
+# License: MIT
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""FallbackMetricFetcher implementation that uses formula generator."""
+
+from frequenz.channels import Receiver
+
+from ... import Sample
+from ..._quantities import QuantityT
+from .. import FormulaEngine
+from .._formula_steps import FallbackMetricFetcher
+from ._formula_generator import FormulaGenerator
+
+
+# This is done as a separate module to avoid circular imports.
+class FallbackFormulaMetricFetcher(FallbackMetricFetcher[QuantityT]):
+    """A metric fetcher that uses a formula generator.
+
+    The formula engine is generated lazily, meaning it is created only when
+    the `start` or `fetch_next` method is called for the first time.
+    Once the formula engine is initialized, it subscribes to its components
+    and begins calculating and sending the formula results.
+    """
+
+    def __init__(self, formula_generator: FormulaGenerator[QuantityT]):
+        """Create a `FallbackFormulaMetricFetcher` instance.
+
+        Args:
+            formula_generator: A formula generator that generates
+                a formula engine with fallback components.
+        """
+        super().__init__(name=formula_generator.namespace)
+        self._formula_generator: FormulaGenerator[QuantityT] = formula_generator
+        self._formula_engine: FormulaEngine[QuantityT] | None = None
+        self._receiver: Receiver[Sample[QuantityT]] | None = None
+
+    @property
+    def is_running(self) -> bool:
+        """Check whether the formula engine is running."""
+        return self._receiver is not None
+
+    def start(self) -> None:
+        """Initialize the formula engine and start fetching samples."""
+        self._formula_engine = self._formula_generator.generate()  # type: ignore[assignment]
+        assert isinstance(self._formula_engine, FormulaEngine)
```

Comment:
**Generics hell** 
 
`generate` method definition:
```
    @abstractmethod
    def generate(
        self,
    ) -> FormulaEngine[QuantityT] | FormulaEngine3Phase[QuantityT]:
        """Generate a formula engine, based on the component graph."""
```

It can return `FormulaEngine` or `FormulaEngine3Phase`. 
* FormulaEngine returns formula output in type `Sample`
* FormulaEngine3Phase returns output as in type `Sample3Phase`

But then `MetricFetcher` works only on FormulaEngine - because it stores receiver that returns `Sample`
FallbackMetricFetcher also has to work on  FormulaEngine - because its output is returned by MetricFetcher.

To solve it in `proper` way we would need separate base class for FormulaEngine and FormulaEngine3Phase.

Reply:
I see, yeah, I also bumped into these issues in #822.

Maybe then the `ignore` won't be needed if you do this instead?

```py
        engine = self._formula_generator.generate()
        # We need this assert because generate() can return a FormulaEngine
        # or FormulaEngine3Phase, but in this case we know it will return a
        # FormulaEngine. This helps to silence `mypy` and also to verify our
        # assumptions are still true at runtime
        assert isinstance(engine, FormulaEngine)
        self._formula_engine = engine
```

Reply:
Maybe nicer solution is to have `generate` and `generate_3_phase`, and different formulas implement different functions.  But not for this PR.

Reply:
@ela-kotulska-frequenz maybe you can add a link to the issue in the comment too :innocent: